lwt/nbd_lwt_unix.ml

```diff
@@ -74,21 +87,28 @@ let tls_channel_of_fd fd role () =
     ignore (Lwt_ssl.ssl_shutdown sock);
     Lwt_ssl.close sock in
 
-  return { read_tls; write_tls; close_tls }
+  return {
+    read_tls = impatient timeout_seconds read_tls;
+    write_tls = impatient timeout_seconds write_tls;
+    close_tls
+  }
 
 
-let cleartext_channel_of_fd fd role_opt =
+let cleartext_channel_of_fd ?(timeout_seconds=0.0) fd role_opt =
   let read_clear = Lwt_cstruct.(complete (read fd)) in
   let write_clear = Lwt_cstruct.(complete (write fd)) in
   let close_clear () = Lwt_unix.close fd in
   let make_tls_channel = match role_opt with
     | None -> None
-    | Some role -> Some (tls_channel_of_fd fd role)
```

Comment:
Hmm, it seems that in this case both the cleartext and the TLS channel that wraps it will have a timeout. I don't think that's an issue though, because if one of them times out, the other timeout will be ignored, because `Lwt.pick` cancels the other thread, and just returns the exception of the first thread that finished.

Reply:
Yes, that's right. But the `read_tls` in the TLS channel does not use (or wrap, or refer to) the `read_clear` of the corresponding cleartext channel. (The same goes for the write functions.)

Once we have a TLS channel, we never call the `read_clear` or `write_clear` functions, so they become irrelevant.

Reply:
Oops true, my bad - I was still thinking in terms of socket options, it's so sad that that solution did not work at all :(